src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/AbstractConditionMapper.java

```diff
@@ -57,36 +59,72 @@
         this.itemMapper = itemMapper;
     }
 
+    public enum UpdateType {
+        SET("SET"),
+        ADD("ADD");
+
+        private final String value;
+
+        UpdateType(String value) {
+            this.value = value;
+        }
+
+        @Override
+        public String toString() {
+            return this.value;
+        }
+    }
+
+    private List<String> getPhysicalUpdateExpressionForAction(RequestWrapper request,
+                                                               Map<String, AttributeValue> actions,
+                                                               String delimiter) {
+        // map virtual field values to physical ones
+        Map<String, AttributeValue> physicalActions = itemMapper.applyForWrite(actions);
+
+        // construct physical update SET expression
+        List<String> clauses = physicalActions.entrySet().stream()
+            .map(entry -> getUpdateClause(request, entry.getKey(), entry.getValue(), delimiter))
+            .collect(Collectors.toList());
+
+        return clauses;
+    }
+
     @Override
     public void applyForUpdate(UpdateItemRequest updateItemRequest) {
         if (updateItemRequest.getUpdateExpression() != null) {
             RequestWrapper request = new UpdateExpressionRequestWrapper(updateItemRequest);
 
             // parse update expression for the field values being set
-            Map<String, AttributeValue> virtualSetActions = parseUpdateExpression(request,
-                updateItemRequest.getConditionExpression());
-            checkArgument(!virtualSetActions.isEmpty(), "Update expression needs at least one SET action");
+            List<? extends Map<String, AttributeValue>> virtualActions = parseUpdateExpression(
+                request, updateItemRequest.getConditionExpression());
+
+            Map<String, AttributeValue> virtualSetActions = virtualActions.get(0);
+            Map<String, AttributeValue> virtualAddActions = virtualActions.get(1);
+
+            checkArgument(!virtualSetActions.isEmpty() || !virtualAddActions.isEmpty(),
+                "Update expression needs at least one supported action: SET or ADD");
 
             // validate no table primary key field is being updated, and that if one field in a secondary index is
             // being updated, then the other is as well
-            validateFieldsCanBeUpdated(virtualSetActions.keySet());
+            validateFieldsCanBeUpdated(virtualSetActions.keySet(), UpdateType.SET);
+            validateFieldsCanBeUpdated(virtualAddActions.keySet(), UpdateType.ADD);
```

Comment:
This validation now seems too restrictive, in that if we have a secondary index on field A and field B, and the update does a SET on A and an ADD on B, this validation will reject the update, when it should be allowed. The reason we have the restriction is that both virtual index field end up being encoded in the same physical attribute, so as long as both fields are part of an update, we should allow it.

Reply:
I think we want to pass in all virtualActions into `validateFieldsCanBeUpdated`, without the update type specified, here. And then in the hash partitioning implementation, call super, and then for the set of ADD actions, validate they are not in any secondary index.